### PR TITLE
Updater services

### PR DESCRIPTION
**Pull request by bbarberBPL**
- Adds `UpdaterService` concern in lib for handling updates
- Moved Transaction specific methods into separate concern for further DRYness
- DRYed out factory service methods to allow them to be shared between factory/updater services
- Implemented updater services for `Institution`, `Collection`, `DigitalObject`, `FileSets` and their associated `Metastreams`. 
- `Added Ability to add and remove child mappings depending on object by passing in `_destroy: '1'` 
- Implemented all PUT/PATCH endpoints associated with each corresponding updater service.
- Only whitelisted certain attributes in an objects strong parameters in controllers

### CODE REVIEW

File: spec/services/curator/collection_updater_service_spec.rb [hunk 1]
```diff
@@ -0,0 +1,47 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe Curator::CollectionUpdaterService, type: :service do
+  before(:all) do
+    @collection ||= create(:curator_collection)
```
Comment (ebenenglish) [inline spec/services/curator/collection_updater_service_spec.rb:7]: Curious about the use of `||=` and `@instance_variable` in these updater specs. Do we actually want objects created in other specs to be used here?

File: app/controllers/curator/metastreams/workflows_controller.rb [hunk 2]
```diff
@@ -24,7 +28,7 @@ def set_workflow
     def workflow_params
       case params[:action]
       when 'update'
-        params.require(:administrative).permit!
+        params.require(:workflow).permit(:publishing_state)
```
Comment (ebenenglish) [inline app/controllers/curator/metastreams/workflows_controller.rb:31]: We want to allow `processing_state` as well here.

File: app/controllers/curator/filestreams/file_sets_controller.rb [hunk 3]
```diff
@@ -35,13 +38,15 @@ def file_set_params
         params.require(:file_set).permit(:ark_id, :created_at, :updated_at,
                                          :file_set_type, :file_name_base, :position,
                                          file_set_of: [:ark_id],
-                                          exemplary_image_of: [:ark_id],
+                                          exemplary_image_of: [:ark_id, :_destroy],
```
[the remark this answers is not in the document]
Comment (bbarberBPL) [reply app/controllers/curator/filestreams/file_sets_controller.rb:41]: it should be in the update only I will correct that

File: app/controllers/curator/digital_objects_controller.rb [hunk 4]
```diff
@@ -41,10 +44,46 @@ def digital_object_params
                                                is_member_of_collection: [:ark_id],
                                                contained_by: [:ark_id],
                                                metastreams: {
-                                                 descriptive: {},
+                                                 descriptive: [:abstract,
```
[the remark this answers is not in the document]
Comment (bbarberBPL) [reply app/controllers/curator/digital_objects_controller.rb:47]: I'll see if i can add some kind of helper method or concern to deal with that

File: spec/controllers/curator/metastreams/descriptives_controller_spec.rb [hunk 5]
```diff
@@ -4,7 +4,6 @@
 require_relative '../shared/shared_formats_and_actions'
 
 RSpec.describe Curator::Metastreams::DescriptivesController, type: :controller do
-  # let(:valid_session) { {} }
   #
   # let(:valid_attributes) {
```
[the remark this answers is not in the document]
Comment (bbarberBPL) [reply spec/controllers/curator/metastreams/descriptives_controller_spec.rb:8]: No I will remove

File: spec/controllers/curator/metastreams/workflows_controller_spec.rb [hunk 6]
```diff
@@ -48,7 +47,15 @@
           }
         end
 
-        include_examples 'shared_formats', include_ark_context: true, has_collection_methods: false, resource_key: 'workflow'
+        let!(:valid_update_attributes) do
```
[the remark this answers is not in the document]
Comment (bbarberBPL) [reply spec/controllers/curator/metastreams/workflows_controller_spec.rb:50]: O I put that in based on the sheet specs. I will change that

File: app/controllers/curator/metastreams/workflows_controller.rb (hunk 2 above)
Comment (bbarberBPL) [reply app/controllers/curator/metastreams/workflows_controller.rb:31]: Will change that

**Comment by bbarberBPL** [comment]
In regards to `_destroy: '1'` I was following rails nested form convention as it usually sets that to '1' rather than true. See [Documentation](https://api.rubyonrails.org/classes/ActiveRecord/NestedAttributes/ClassMethods.html)

If you still think it is worth the change I can do that but let me know your thoughts.

File: app/models/curator/mappings/host_collection.rb [hunk 7]
```diff
@@ -9,6 +9,8 @@ class Mappings::HostCollection < ApplicationRecord
     has_many :desc_host_collections, inverse_of: :host_collection,
              class_name: 'Curator::Mappings::DescHostCollection', dependent: :destroy
 
+    scope :name_lower, ->(name) { where('lower(name) = ?', name.downcase) }
```
[the remark this answers is not in the document]
Comment (bbarberBPL) [reply app/models/curator/mappings/host_collection.rb:12]: It helps with the prevention of duplicates in the database. Used in `digital_object_factory.rb#find_or_create_host_collection`

File: app/models/curator/metastreams/workflow.rb [hunk 8]
```diff
@@ -8,6 +8,7 @@ class Metastreams::Workflow < ApplicationRecord
     enum processing_state: { derivatives: 0, complete: 1 }.freeze
 
     validates :ingest_origin, presence: true
+    validates :publishing_state, presence: true
```
[the remark this answers is not in the document]
Comment (bbarberBPL) [reply app/models/curator/metastreams/workflow.rb:11]: Forgot to remove that when i was writing the specs for this. Will take that out

File: app/services/concerns/curator/mappings/create_or_replace_exemplary.rb [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+      include InstanceMethods
+    end
+
+    module InstanceMethods
+      protected
+
+      def create_or_replace_exemplary!(exemplary_file_set_ark_id = nil)
+        return if exemplary_file_set_ark_id.blank?
+
+        exemplary_file_set = @record.file_sets.exemplaryable.find_by!(ark_id: exemplary_file_set_ark_id)
```
[the remark this answers is not in the document]
Comment (bbarberBPL) [reply app/services/concerns/curator/mappings/create_or_replace_exemplary.rb:17]: `exemplaryable` is a special scope that is used to find file sets that are available to be an exemplary image for the specific exemplary object(Collection/DigitalObject). queried by FileSet Types that have `exemplary_image_of` defined and based on the `file_sets` relationships for the exemplary object. It's more of a utility/helper scope if anything if that makes sense.

File: app/services/concerns/curator/mappings/create_or_replace_exemplary.rb [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+        @record.build_exemplary_image_mapping if @record.exemplary_image_mapping.blank?
```
[the remark this answers is not in the document]
Comment (bbarberBPL) [reply app/services/concerns/curator/mappings/create_or_replace_exemplary.rb:18]: Its a rails method that gets defined automatically.

File: app/services/curator/digital_object_updater_service.rb [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+    end
+
+    private
+
+    def member_collections_scope
+      # NOTE: the #admin_set should always be mapped
+      Curator.collection_class
+                              .select(:id, :ark_id, :institution_id)
+                              .where(institution_id: @record.institution.id)
+                              .where
```
[the remark this answers is not in the document]
Comment (bbarberBPL) [reply app/services/curator/digital_object_updater_service.rb:71]: no Its its chained to a `not` which basically tells it to give me the records that aren't in list of identifiers

File: app/services/curator/metastreams/descriptive_updater_service.rb [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+        @record.save!
+      end
+
+      return @success, @result
+    end
+
+    protected
+
+    def term_mappings_update!
+      needs_removal = proc { |term_attrs| term_attrs.key?(:_destroy) && term_attrs[:_destroy] == '1' }
```
[the remark this answers is not in the document]
Comment (bbarberBPL) [reply app/services/curator/metastreams/descriptive_updater_service.rb:70]: I can add a Constant proc to the `Services::UpdaterService` to add the ability to check for that in all instances of an updater class.

File: spec/services/curator/collection_updater_service_spec.rb (hunk 1 above)
Comment (bbarberBPL) [reply spec/services/curator/collection_updater_service_spec.rb:7]: Not sure if they get shared between specs if you memonize them this way. wanted to ensure the data generated for this set of specs didn't change in any way.

**Comment by ebenenglish** [comment]
> In regards to `_destroy: '1'` I was following rails nested form convention as it usually sets that to '1' rather than true.

OK, that's fine.

File: app/services/curator/metastreams/descriptive_updater_service.rb (hunk 12 above)
[the remark this answers is not in the document]
Comment (ebenenglish) [reply app/services/curator/metastreams/descriptive_updater_service.rb:70]: Just to be clear, I was just thinking that `term_attrs.key?(:_destroy) && term_attrs[:_destroy] == '1'` is redundant in that `term_attrs[:_destroy] == '1'` would work fine on its own. If the `:_destroy` key is missing, then `term_attrs[:_destroy]` will eval to `nil`.

[the remark this answers is not in the document]
Comment (bbarberBPL) [reply app/services/curator/metastreams/descriptive_updater_service.rb:70]: This is true. I can correct that. And also DRY it out a little bit.

File: app/models/curator/collection.rb [hunk 13]
```diff
@@ -17,6 +17,22 @@ class Collection < ApplicationRecord
 
     has_many :collection_members, inverse_of: :collection, class_name: 'Curator::Mappings::CollectionMember', dependent: :destroy
 
+    # NOTE: Should this include the file sets included by the collection members as well?
```
[the remark this answers is not in the document]
Comment (bbarberBPL) [reply app/models/curator/collection.rb:20]: This is what we talk about a little while ago in teams. I removed that comment. so you can resolve this convo

File: app/controllers/curator/filestreams/file_sets_controller.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (ebenenglish) [reply app/controllers/curator/filestreams/file_sets_controller.rb:41]: Still needs to be fixed.

File: spec/controllers/curator/metastreams/workflows_controller_spec.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (ebenenglish) [reply spec/controllers/curator/metastreams/workflows_controller_spec.rb:50]: `published` is not a valid value for `processing_state` -- it should be `derivatives` or `complete`

File: app/controllers/concerns/curator/descriptive_params.rb [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+        note: [:label, :type],
+        physical_location: [:authority_code, :id_from_auth, :label, :affiliation, :name_type],
+        publication: [:edition_name, :edition_number, :volume, :issue_number],
+        related: [:constituent, referenced_by_url: [], references_url: [], other_format: [], review_url: []],
+        resource_types: [:label, :authority_code, :id_from_auth],
+        subject: [topics: [:label, :authority_code, :id_from_auth], names: [:label, :name_type, :authority_code, :id_from_auth, :affiliation], geos: [:label, :authority_code, :id_from_auth, :coordinates, :bounding_box, :area_type], titles: [:label, :id_from_auth, :subtitle, :authority_code, :display, :display_label, :usage, :supplied, :language, :type, :part_number, :part_name], temporals: [], dates: []],
+        title: [primary: [:label, :id_from_auth, :authority_code, :subtitle, :display, :display_label, :usage, :supplied, :language, :type, :part_number, :part_name], other: [:label, :id_from_auth, :subtitle, :authority_code, :display, :display_label, :usage, :supplied, :language, :type, :part_number, :part_name]]
+      ].freeze
+
+      def decriptive_permitted_params(operation = 'create')
```
Comment (ebenenglish) [inline app/controllers/concerns/curator/descriptive_params.rb:55]: Should be `descriptive_permitted_params`

**Pull request merged**